csp/random.py

```diff
@@ -0,0 +1,167 @@
+        value: The value to tick when there are events (similar to csp.timer)
+    """
+    with csp.alarms():
+        event = csp.alarm(bool)
+
+    with csp.state():
+        s_rng = np.random.default_rng(seed)
+        s_scheduled_event = None
+
+    if csp.ticked(rate):
+        if rate < 0:
+            raise ValueError(f"{csp.now()}: rate must be non-negative")
+        if s_scheduled_event:
+            csp.cancel_alarm(event, s_scheduled_event)
+        if rate > 0:
+            seconds = s_rng.exponential(1.0 / rate)
+            s_scheduled_event = csp.schedule_alarm(event, timedelta(seconds=seconds), True)
+
+    if csp.ticked(event):
+        seconds = s_rng.exponential(1.0 / rate)
+        s_scheduled_event = csp.schedule_alarm(event, timedelta(seconds=seconds), True)
+        return value
+
+
+def _make_brownian_increment(t, s_rng, drift, s_cov_decomp):
+    # Make the brownian increment as efficiently as possible
+    values = s_rng.normal(scale=np.sqrt(t), size=drift.size)
+    np.dot(s_cov_decomp, values, out=values)
+    np.add(drift * t, values, out=values)
+    return values
+
+
+def _matrix_decomposition(matrix, now):
+    # Split matrix decomposition into its own function to help with profiling and make it easier to replace
+    # We use SVD instead of Cholesky because it's more stable and handles the zero variance case.
+    U, S, Vt = np.linalg.svd(matrix)
+    if not np.allclose(matrix, matrix.T):
+        raise ValueError(f"{now}: covariance not symmetric")
+    if not np.allclose(U, Vt):
+        raise ValueError(f"{now}: covariance not positive semidefinite")
+    return np.dot(U, np.sqrt(np.diag(S)), out=U)
+
+
+@csp.node
+def brownian_increments(
+    trigger: ts[object],
+    drift: ts[Numpy1DArray[float]],
+    covariance: ts[NumpyNDArray[float]],
+    seed: object,
+) -> ts[Numpy1DArray[float]]:
+    """Generate multi-dimensional Brownian increments at the trigger times, with time-varying drift and covariance.
+
+    Args:
+        trigger: When to return the value of the process
+        drift: Drift parameter (per second), i.e. array of length n
+        covariance: Covariance matrix (per second), i.e. array of size nxn
+        seed: The seed for the numpy random Generator. Can be anything accepted by np.random.default_rng
+    """
+    with csp.state():
+        s_rng = np.random.default_rng(seed)
+        s_cov_decomp = None  # Placeholder for covariance matrix decomposition
+        s_last_change = None  # Placeholder for last csp.now()
+        s_last_drift = None  # Placeholder for last value of drift
+        s_last_value = None  # Placeholder for cumulative value between trigger ticks
+
+    if csp.ticked(drift, covariance) and csp.valid(drift, covariance):
+        if s_last_change is None:
+            if not drift.ndim == 1:
+                raise ValueError(f"{csp.now()}: drift must be 1-dimensional")
+            if not covariance.ndim == 2:
```

Comment:
I would combine the 2-dimensional, square and symmetric check into one single check which reports "covariance must be a symmetric matrix" (symmetric being a subset of square which is subset of 2D).

Reply:
Good catch, I will do this.